builtin/provisioners/puppet/resource_provisioner.go

```diff
@@ -0,0 +1,337 @@
+func Provisioner() terraform.ResourceProvisioner {
+	return &schema.Provisioner{
+		Schema: map[string]*schema.Schema{
+			"server": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"server_user": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "root",
+			},
+			"os_type": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"nix", "windows"}, false),
+			},
+			"use_sudo": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"autosign": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"open_source": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"certname": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"pp_role": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"environment": &schema.Schema{
+				Type:     schema.TypeString,
+				Default:  "production",
+				Optional: true,
+			},
+		},
+		ApplyFunc:    applyFn,
+		ValidateFunc: validateFn,
+	}
+}
+
+func applyFn(ctx context.Context) (rerr error) {
+	output := ctx.Value(schema.ProvOutputKey).(terraform.UIOutput)
+	state := ctx.Value(schema.ProvRawStateKey).(*terraform.InstanceState)
+	configData := ctx.Value(schema.ProvConfigDataKey).(*schema.ResourceData)
+
+	p, err := decodeConfig(configData)
+	if err != nil {
+		return err
+	}
+
+	p.instanceState = state
+	p.output = output
+
+	if p.OSType == "" {
+		switch connType := state.Ephemeral.ConnInfo["type"]; connType {
+		case "ssh", "":
+			p.OSType = "nix"
+		case "winrm":
+			p.OSType = "windows"
+		default:
+			return fmt.Errorf("Unsupported connection type: %s", connType)
+		}
+	}
+
+	switch p.OSType {
+	case "nix":
+		p.runPuppetAgent = p.nixRunPuppetAgent
+		p.installPuppetAgent = p.nixInstallPuppetAgent
+		p.uploadFile = p.nixUploadFile
+		p.defaultCertname = p.nixDefaultCertname
+		p.isPuppetEnterprise = p.nixIsPuppetEnterprise
+	case "windows":
+		p.runPuppetAgent = p.windowsRunPuppetAgent
+		p.installPuppetAgent = p.windowsInstallPuppetAgent
+		p.uploadFile = p.windowsUploadFile
+		p.UseSudo = false
+		p.defaultCertname = p.windowsDefaultCertname
+		p.isPuppetEnterprise = p.windowsIsPuppetEnterprise
+	default:
+		return fmt.Errorf("Unsupported OS type: %s", p.OSType)
+	}
+
+	comm, err := communicator.New(state)
+	if err != nil {
+		return err
+	}
+
+	retryCtx, cancel := context.WithTimeout(ctx, comm.Timeout())
+	defer cancel()
+
+	err = communicator.Retry(retryCtx, func() error {
+		return comm.Connect(output)
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := comm.Disconnect(); err != nil {
+			rerr = err
+		}
+	}()
+
+	p.comm = comm
+
+	_, ok := configData.GetOkExists("open_source")
+	if !ok {
+		isPE, err := p.isPuppetEnterprise()
```

Comment:
I don't quite follow why this is a remote command instead of a required configuration attribute - could you help me explain the thought process here? I'd expect users would  explicitly declare enterprise vs open source.

Reply:
My thought here was to simplify the user experience so that the users didn't have to specify this and the same provisioner block could be reused regardless of whether connecting to an open source or enterprise master. There's still the option of explicitly configuring this if the user wanted or the automatic determination fails for some reason.

Reply:
OK, I've removed that bit of code